src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs

```diff
@@ -650,29 +483,27 @@ public bool IsEmpty
         {
             get
             {
-                if (_headList == null)
-                    return true;
-
-                bool lockTaken = false;
-                try
+                if (_headList != null)
                 {
-                    FreezeBag(ref lockTaken);
-                    ThreadLocalList currentList = _headList;
-                    while (currentList != null)
+                    bool lockTaken = false;
+                    try
                     {
-                        if (currentList._head != null)
-                        //at least this list is not empty, we return false
+                        FreezeBag(ref lockTaken);
```

Comment:
It'll only be false if an exception is getting thrown, e.g. an assert for true after this call would always pass.